### ISSUE TITLE
Suporting cursors(get one row at a time)

### ISSUE THREAD

**Issue by derkan** [samonzeweb/godb#6]
Hi Sam,

It will be helpful to expose a function from `SelectStatement` to get rows one by one. It will be useful if user doesn't want to fetch all data from DB. 

Of course we can use `LIMIT` and `OFFSET` but, if size of row is big(for example: rows with columns containing file contents) it will be helpful to get one by one and `break` if necessary.

derkan.

**Comment by samonzeweb** [comment]
Hi Derkan,

I agree, it could be useful for long batch processing.

But before add such a feature I need to check what could goes wrong while iterating on the resulset. There are two parts :

* is the godb API "safe" (could the user do something wrong accidentally).
* is godb safe for itself.

For the second part I think the prepared statement cache need some works. I'll also need to check the transactions part, just to be sure. But nothing should prohibit this feature development, it's just not trivial.

Really a good feature to add. I'll work on it sooner or later.
Thanks.

**Comment by derkan** [comment]
Hi Sam,
upper.io db has a nice implementation which is used like this:
```go
 q := sess.Select("author").From("books").Where("id > ?",10)
 iter := q.Iterator()
 defer iter.Close()
 for iter.Next() {
 var authorName string
 if err := iter.Scan(&authorName); err != nil {
	log.Fatalf("iter err :%v",err)
 }
 log.Printf("iter %v", authorName)
```

**Comment by samonzeweb** [comment]
I see two way to do it : 

* An iterator 
* A function with take another function (a closure) called for each row

I prefer the second, but the first offer more flexibility. And of course with an iterator it will be easy to do the second. Then I agree that an iterator is the way to go.

**Comment by samonzeweb** [comment]
I worked on an iterator, at least with the select statement.

An example can be found in tests (in the  DoAndScanOneByOne branch) : https://github.com/samonzeweb/godb/blob/DoAndScanOneByOne/dbtests/common/statements_tests.go#L154

I've struggled with nested iterator with transactions. Without transactions the sql package uses different connexions, but with a transaction it executes the requests on the same connexion, and it causes troubles with PostgreSQL, MySQL and SQL Server drivers. To avoid troubles I decided to never use an existing transaction. It's the safest choice I could do now.
